### PR TITLE
Modularizing and clarifying core engine

### PR DESCRIPTION
**Pull request by mramendi**

### CODE REVIEW

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+This strategy is similar to what was available by default in jBPM version 5.x. 
+endif::DROOLS,JBPM,OP[]
+Use this strategy for simple use cases.
+
+It has the following characteristics that can be important for a usage scenario:
+
+* It has a small memory footprint - single instance of the runtime engine and the task service.
+* It is simple and compact in design and usage.
+* It is a good fit for low to medium load on the {PROCESS_ENGINE} because of synchronized access.
+* Because of the single KieSession instance, all state objects (such as facts) are directly visible to all process instances and vice versa.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:18]: Suggest rewording for parallelism in this list, such as "It provides all state objects to all process instances, due to the single `KieSession` instances." Or something.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+* The strategy is not contextual. When you retrieve instances of RuntimeEngine from a singleton RuntimeManager, you do not need to take the `Context` instance into account. Usually, you can use EmptyContext.get() as the context, although a null argument is acceptable as well
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:19]: Again for parallelism, "It is not contextual."  Also, in this sentence need to unicode `RuntimeEngine`, `RuntimeManager` and `EmptyContext.get()`.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+* The runtime manager keeps track of the ID of the `KieSession`, so that the same session remains in use after a `RuntimeManager` restart. The ID is stored as a serialized file in a temporary location in the file system that, depending on the environment, can be one of the following directories:
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:20]: Again for parallelism, "It instructs the `RuntimeManager` class to keep track of ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+** The value of the `jbpm.data.dir` system property
+** The value of the `jboss.server.data.dir` system property
+** The value of the `java.io.tmpdir` system property
+
+[WARNING]
+====
+A combination of the Singleton strategy and the EJB Timer Scheduler has a limitation that it might raise Hibernate issues under load. The combination is not recommended for production use. The EJB Timer Scheduler is the default in the {KIE_SERVER}.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:27]: For minimalism "Scheduler might raise Hibernate issues under load."
Also "Scheduler is the default scheduler in ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+====
+
+*Per request strategy*: instructs `RuntimeManager` to provide a new instance of `RuntimeEngine` for every request. One or more invocations of the {PROCESS_ENGINE} within a single transaction are considered a single request.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:30]: Same formatting suggestion as above, and suggest "This strategy instructs ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+The same instance of `RuntimeEngine` must be used within single transaction to ensure correctness of state. Otherwise, an operation completed in one call would not be visible in the next call.
+
+This stragegy is "stateless", as process state is preserved only within the request. When a request is completed, the `RuntimeEngine` is permanently destroyed. If persistence is used, information related tothe KIE session is removed from the persistence database as well.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:34]: Change to "strategy" (typo), also just "is stateless" (no quotes). We don't do quotes in documentation for emphasis. Use italics if needed.

Also "to the" to fix typo

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+The strategy has the following characteristics:
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:36]: Suggest "This strategy" for consistency.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+* It provides completely isolated {PROCESS_ENGINE} and task service operations for every request.
+* It is completely stateless, storage of any facts happens only for the duration of the request.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:39]: Suggest "stateless, where facts are stored only for the duration ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+* It is a good fit for high-load, stateless processes, where no facts or timers must be preserved between requests.
+* The `KieSession` is only available during the lifetime of a request and is destroyed at the end of the request.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:41]: Again for parallelism, "It uses a `KieSession` instance that is only ..."

Also suggest "during the life of"

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+* The strategy is not contextual. When you retrieve instances of RuntimeEngine from a singleton RuntimeManager, you do not need to take the `Context` instance into account. Usually, you can use EmptyContext.get() as the context, although a null argument is acceptable as well
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:42]: Again, "It is not contextual ..." And again with the unicode formatting here, mentioned in the earlier comment (`RuntimeManager`, etc.)

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+*Per process instance strategy*: instructs `RuntimeManager` to maintain a strict relationship between a KIE session and a process instance. Each `KieSession` is available as long as the `ProcessInstance` to which it belongs is active.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:44]: Same formatting suggestion as the other two, and suggest "This strategy instructs ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+This strategy provides the most flexible approach to use advanced capabilities of the {PROCESS_ENGINE} like rule evaluation and isolation between process instances. It maximizes performance and reduces potential bottlenecks introduced by synchronization. At the same time, unlike the request strategy, it reduces the number of KIE sessions to the actual number of process instances, rather than the total number of requests.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:46]: Suggest "the most flexibility with advanced capabilities ..."

Also "unlike the per request strategy ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+This strategy has the following characteristics:
+
+* It provides isolation for every process instance.
+* It maintains a strict relationship between `KieSession` and `ProcessInstance` to ensure that it will always deliver the same `KieSession` for a given `ProcessInstance`.
+* It merges the lifecycle of `KieSession` with `ProcessInstance`, and both are disposed when the process instance completes or aborts.
+* It enables maintaining data, such facts and timers, in the scope of the process instance. Only the process instance has access to the data.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:53]: Suggest "It enables maintenance of data, such as facts ..."  (note typo)  Just sounds off to me otherwise.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+* It introduces some overhead because of the need to look up and load the `KieSession` for the process instance.
+* It validates every usage of a `KieSession` so it cannot be used for other process instances. An exception is thrown if another process instance would use the same `KieSession`.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:55]: Probably should be "if another process instance uses the same ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+* The strategy is contextual and accepts the following context instances:
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:56]: Again for parallelism, "It is contextual and ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+** `EmptyContext` or null: When starting a process instance, as there is no process instance id available yet
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:58]: Change to "Used when starting a process instance because no process ID is available yet"  Note cap ID. Also note that per IBM SG, avoid empty "there is/are".

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+** `ProcessInstanceIdContext`: Used after process instance was created
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:59]: "Used after the process ..."

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+** `CorrelationKeyContext`: Used as an alternative to `ProcessInstanceIdContext` to use a custom (business) key instead of the process instance id
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:60]: Again "instance ID" (cap)

**Review by sterobin — COMMENTED** [review]
@mramendi First installment of comments, for sections 59 and 59.1. Will pick up with 59.2 and on next week. 

As a comment for future PRs on this doc and really any doc, I recommend a closer personal review before sending for review. My review is being slowed down by typos, formatting issues, and inconsistencies that imo should be caught with a proper close proofread before final peer review. This is especially important when converting community content, which naturally follows looser standards. Will ping you with my next installment hopefully Monday.

On a more positive note, it's thrilling to finally see this stuff taking shape for enterprise! Been a long time waiting. Thank you SO much for all your work through this Misha. I can completely empathize, as I've done it many times now in our doc set.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/service-quartz-con.adoc [hunk 19]
```diff
@@ -0,0 +1,49 @@
+[id='service-quartz-con_{context}']
+= Quartz-based Timer Service
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/service-quartz-con.adoc:2]: Lower case "timer service" I assume.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/service-query-con.adoc [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+* Runtime operations: 
+** simple query based on `QueryParam` as filter provider
+** advanced query based on `QueryParamBuilder` as filter provider
+
+DashBuilder DataSets provide support for multiple data sources, such as CSV, SQL, and Elastic Search. However, the {PROCESS_ENGINE} uses a RDBMS based backend and focuses on SQL-based data sets.
+
+Therefore, the {PROCESS_ENGINE} query service is a subset of DashBuilder DataSets capabilities that enables efficient queries with a simple API.
+
+[float]
+== Key classes of the query service
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/service-query-con.adoc:25]: Because this module file is already level offset of 3, these subsections are pushed to level 4 and level 5 offset, which makes it render very poorly. We try to stick to 3 levels of offset max, with exceptions for 4. I don't think we want 5 ever. However, this is a genuine problem in complex documents like this with many aspects of many pieces.

That said, to avoid excessive offset, we can just avoid unnecessary titles when practical. In this case, I think the titles are unnecessary anyway and so an easy fix, imo.

I'd just nix the titles and run the text as follows:

```
... queries with a simple API.

The query service uses the following key classes:
...

The `QueryResultMapper` class maps ...
...

The `QueryParamBuilder` class provides ...
...
```

(Also note that the "Key classes" title would've needed a lead-in sentence anyway.)

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/deployment-versions-con.adoc [hunk 21]
```diff
@@ -0,0 +1,72 @@
+[id='deployment-versions-con_{context}']
+= Deployments and different process versions
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/deployment-versions-con.adoc:2]: The sub-headings are in level 4 offset, but I don't think we can rework that unless we ditch this top-level module and make the sub-sections the modules, which has its own downside. So maybe stuck with it in this case. Up to you.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/deployment-versions-con.adoc [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+The deployment service puts business assets into an execution environment. However, in some cases additional management is required to make the assets available in the right context. Notably, if you deploy several versions of the same process, you must ensure that process instances use the correct version.
+
+[float]
+== Activation and Deactivation of deployments
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/deployment-versions-con.adoc:7]: Change to "and deactivation" (lowercase)

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/deployment-versions-con.adoc [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+To enable this scenario, use the following methods of the deployment service:
+
+* `activate`: activates a given deployment so it can be available for interaction. You can lost its process definitions and start new process instances for this deployment.
+
+* `deactivate`: deactivates a deployment. Disables the option to list process definitions and to start new process instances of processes in the deployment. However, you can continue working with the process instances that are already active, for example, signal events and interact with user tasks.
+
+You can use this feature for smooth transition between project versions without the need for process instance migration.
+
+[float]
+== Invoking latest version of  processes
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/deployment-versions-con.adoc:22]: Suggest "Invocation of the latest version of a process" (non-proc, and for consistency)

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/deployment-synchronization-con.adoc [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+When embedded in jbpm-console/kie-wb, they utilized the sistem.git VFS repository to preserve deployed units across server restarts.
+While that works fine, it comes with some drawbacks:
+
+* not available for custom systems that use services
+* requires complex setup in cluster - zookeeper and helix
+
+With version 6.2 jbpm services come with a deployment synchronizer
+endif::DROOLS,JBPM,OP[]
+ifdef::PAM,DM[]
+{PROCESS_ENGINE} services include a deployment synchronizer
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/deployment-synchronization-con.adoc:15]: Change to {PROCESS_ENGINE_CAP} if begins sentence. But it's probably good to try to avoid this cap variable anyway, unless you really need it. Issue is that for community it becomes "JBPM engine" (instead of "jBPM engine"), which isn't terrible, but is more for titles when you can't avoid it. If you can avoid it in prose, by using "The {PROCESS_ENGINE}" or whatever, probably best.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/threads-process-con.adoc [hunk 25]
```diff
@@ -0,0 +1,26 @@
+[id='threads-process-con_{context}']
+// Stetson: shall we condition this text, or a part of it, only for community documentation?
+= Threads in the {PROCESS_ENGINE}
+
+We can refer to two types of "multi-threading":  _logical_ and _technical_. _Technical multi-threading_ involves multiple threads or processes that are started, for example, by a Java or  C program. _Logical multi-threading_ happens in a BPM process, for example, after the process reaches a parallel gateway. In execution logic, the original process splits into two processes that run in a parallel fashion.
```
Comment (sterobin) [inline doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/threads-process-con.adoc:5]: Informal and unclear, suggest something like, "Applications typically implement two general types of multi-threading: ..."  Also note again that we shouldn't be using quotes for emphasis in docs.

**Review by sterobin — APPROVED** [review]
@mramendi Done for now. As we discussed in chat, it seems you're getting these PRs out in more rough state and are planning to do a final close proof read and polishing in the end. So for the remainder of this PR and for future PRs, I'm just doing a quick high-level look for anything glaring at the structural, modular, or flow level. Will do an official peer review of the whole of it after you've polished it in the end.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/kiesession-con.adoc [hunk 26]
```diff
@@ -0,0 +1,13 @@
+[id='kiesession-con-{context}']
+= KIE session
```
[the remark this answers is not in the document]
Comment (mramendi) [reply doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/kiesession-con.adoc:2]: Not processed as I do not think the difference exists - saved in backlog doc

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/event-listeners-con.adoc [hunk 27]
```diff
@@ -0,0 +1,57 @@
+[id='event-listeners-con_{context}']
+= Event Listeners
+
+You can develoip a class that implements the `ProcessEventListener`. This class can listen to process-related events, like starting or completing a process, entering and leaving a node, and so on.
```
[the remark this answers is not in the document]
Comment (mramendi) [reply doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/event-listeners-con.adoc]: I moved event listeners out of the kie base/session section and into their own section like threads, because it turns out they are used a lot when using services.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/event-listeners-con.adoc [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+This approach reflects cause and effect  relationships between events. However, the timing and order of "after" event calls does not correspond to an "intuitive" meaning. For example, an `afterProcessStarted` call can happen after the `afterNodeLeft` calls for some nodes in the process.
+
+In general, to be notified when a particular event occurs, use the "before" call for the event. Use an "afrer" call only if you want to make sure that all processing related to this event has ended, for example, when you want to be notified when all steps associated with starting a particular process instance have been completed.
+
+Depending on the type of node, some nodes might only generate "node left" calls and others might only generate "node triggered" calls. For example, "catch intermediate event" nodes do not generate "node triggered" calls, because they are not triggered by another process node. Similarly, "throw intermediate events" nodes do not generate "node left" calls, as these nodes do not have an outgoing connection to another node.
```
[the remark this answers is not in the document]
Comment (mramendi) [reply doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/event-listeners-con.adoc:43]: https://access.redhat.com/documentation/en-us/red_hat_process_automation_manager/7.8/html/process_designer_business_process_model_and_notation_bpmn2_reference_guide/bpmn-events-con#bpmn-intermediate-events-ref_bpmn-reference is where product docs describe the nodes now. We could do without quotes as "nodes that catch an intermediate event" if that would be better - I don't see it as better but this is not a hill to die on.

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc [hunk 29]
```diff
@@ -0,0 +1,60 @@
+[id='runtime-manager-strategies-con_{context}']
+= Runtime manager strategies
```
[the remark this answers is not in the document]
Comment (mramendi) [reply doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:2]: I really don't want to switch to the class every time. "The runtime manager" is more than that one class, it is a level of control over the system, intermediate between "services" and "base/session".

File: doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc (hunk 2 above)
Comment (mramendi) [reply doc-content/jbpm-docs/src/main/asciidoc/CoreEngine/runtime-manager-strategies-con.adoc:19]: I did not find perfect parallelism possible without significant contortion of the language, however I made sure all lines have either "it" or "this strategy/the strategy".
